### DISCUSSION TITLE
Motion Gen doesn't generate solution when enable_opt is set to TRUE and enable_graph is set to FALSE

### DISCUSSION THREAD

**Discussion by sanjkar126** [NVlabs/curobo#151] [Software Q&A]
Hi,

I am using Motion Planning using curobo motion_gen. Unfortunately I am facing some issues with my configuration.

- **Issue#1** : When I try to plan with obstacles motion_gen fails to find a solution even though there is one. I am able to confirm this because if I set enable_graph to TRUE and set enable_opt to FALSE then it finds a path where as if I change it the other way then it fails.

- **Issue#2**: With just enable_graph as TRUE; I can observe that the Path generated doesn't avoid the obstacles. This isn't consistent but I would say it happens 2 out of 10 times.

My configuration:

```
### Initialization ###
motion_gen_config = MotionGenConfig.load_from_robot_config(
            "my_robot.yml",
            WorldConfig(), # Empty World
            self.tensor_args,
            collision_checker_type=CollisionCheckerType.MESH,
            num_trajopt_seeds=12, 
            num_graph_seeds=12,
            use_cuda_graph=True,
            interpolation_dt=0.01, 
            maximum_trajectory_time=1.0,
            collision_cache={"obb": n_obstacle_cuboids, "mesh": n_obstacle_mesh},
            trajopt_dt=trajopt_dt,
            trajopt_tsteps=trajopt_tsteps,
            trim_steps=trim_steps,
            collision_activation_distance=0.0, # Not sure if this is needed
            interpolation_steps=35, # Had to set this or else it was generating 5000 Trajectory points
        )

 self.motion_gen = MotionGen(motion_gen_config)
 print("warming up...")
 self.motion_gen.warmup(enable_graph=False, warmup_js_trajopt=False)
  
### World Update ### (Obstacles are passed as ros2 Markers)
world_objs = WorldConfig.from_dict(
            {
                "cuboid": obstacle_dict
            }
        )

world_obj_mesh = world_objs.get_mesh_world()

world_cfg = WorldConfig(cuboid=world_objs.cuboid, mesh=world_obj_mesh)
self.motion_gen.clear_world_cache()
self.motion_gen.update_world(world_cfg)
  
 ### Planning ###       
 plan_config = MotionGenPlanConfig(
            enable_opt=False,
            enable_graph=True,
            #enable_graph_attempt=2,
            max_attempts=100,
 )
 result = self.motion_gen.plan_single(start_state, target_pose, plan_config)
        
```

**Comment by balakumar-s** [comment]
What did you set as values for these? 
`collision_cache={"obb": n_obstacle_cuboids, "mesh": n_obstacle_mesh},`

**Reply by sanjkar126** [reply]
`
n_obstacle_cuboids = 30
n_obstacle_mesh = 100
`

**Reply by sanjkar126** [reply]
Hi Balakumar,
Any updates on this?

I do have followup question. I am trying to run benchmark numbers comparing Curobo and MoveIt(OMPL). 

- Experiment is run on Jetson Orin
- Planning for the same scene where I create around 7K obstacles which are small cubes. (For this experiment I set n_obstacle_cuboids = n_obstacle_mesh = 1000000)
- When i run planning. MoveIt plans around 18seconds where as Curobo takes around 25seconds. 

I am surprised as my expectation was that Curobo will outperform MoveIt as it is parallelizing on the GPU. I am guessing that my motion_gen config isn't accurate which is causing significant overhead. (I am also able to see that the GPU is utilized 99% when running the planner with Curobo)

**Reply by balakumar-s** [reply]
1. You need to set interpolation_steps = 5000. Then look at `cmd_plan = result.get_interpolated_plan()` for the solution which will be varying length depending on the trajectory distance.
2. You need to send an obstacle when initializing motion gen instance, you can then update the world model by passing as many obstacles as the cache you setup for. An example world config could be "collision_table.yml"
3. Instead of these lines:
```
world_objs = WorldConfig.from_dict(
            {
                "cuboid": obstacle_dict
            }
        )

world_obj_mesh = world_objs.get_mesh_world()

world_cfg = WorldConfig(cuboid=world_objs.cuboid, mesh=world_obj_mesh)
``` 
use:
```
world_objs = WorldConfig.from_dict(
            {
                "cuboid": obstacle_dict
            }
        )
world_cfg = world_objs.get_collision_check_world()
```
This avoids duplicating of obstacles as cuboids and meshes, instead it converts all obstacles in world_objs to geometries that are supported in cuRobo. 

If you want to convert all obstacles to cuboids, you can call `world_cfg = world_objs.get_obb_world()` instead of `get_collision_check_world()`


If you are only evaluating on cuboids, you can use only the cuboid collision checker by:

```
# from curobo.geom.sdf.world import CollisionCheckerType
MotionGenConfig.load_from_robot_config(...
collision_checker_type=CollisionCheckerType.PRIMITIVE,
collision_cache={"obb": n_obstacle_cuboids},
)

```
If you want to use the mesh collision checker as well, then you need to send a mesh during initialization,

```
world_cfg_table = WorldConfig.from_dict(
  load_yaml(join_path(get_world_configs_path(), "collision_table.yml"))
)
world_cfg_table.cuboid[0].pose[2] -= 0.04
world_cfg1 = WorldConfig.from_dict(
  load_yaml(join_path(get_world_configs_path(), "collision_table.yml"))
).get_mesh_world()
world_cfg1.mesh[0].name += "_mesh"
world_cfg1.mesh[0].pose[2] = -10.5
world_cfg = WorldConfig(cuboid=world_cfg_table.cuboid, mesh=world_cfg1.mesh)

# load this world_cfg as world_file in load_from_robot_config
```

Let us know once you integrate these fixes. This should at least fix your issue of enable_opt vs enable_graph. If it doesn't, then we have to look at your robot configuration file.

**Reply by balakumar-s** [reply]
Regarding Jetson Orin:
1. Make sure you have set it to MAXN mode and also run `sudo jetson_clocks` before running cuRobo.

**Reply by sanjkar126** [reply]
Hello Balakumar,

Thanks for your inputs. I did update things based on your suggestion. Still enable_opt can't find solution. Updates I made:

_### Motion Gen Warm Up ###_
```
#Obstacle Cache Values
n_obstacle_cuboids = n_obstacle_mesh = 10000000
motion_gen_config = MotionGenConfig.load_from_robot_config(
    "my_robot.yml",
    "collision_table.yml",
    self.tensor_args,
    collision_checker_type=CollisionCheckerType.PRIMITIVE,
    num_trajopt_seeds=2,
    num_graph_seeds=2,
    use_cuda_graph=True,
    interpolation_dt=0.01, 
    maximum_trajectory_time=1.0,
    collision_cache={"obb": n_obstacle_cuboids, "mesh": n_obstacle_mesh},
    trajopt_dt=trajopt_dt,
    trajopt_tsteps=trajopt_tsteps,
    trim_steps=trim_steps,
    collision_activation_distance=0.0,
    interpolation_steps=5000,
)

self.motion_gen = MotionGen(motion_gen_config)   
print("warming up...")
self.motion_gen.warmup(enable_graph=False, warmup_js_trajopt=False, parallel_finetune=True)
```

_### Updating Obstacles ###_
```
world_objs = WorldConfig.from_dict(
            {
                "cuboid": obstacle_dict
            }
        )
world_cfg = world_objs.get_collision_check_world()
motion_gen.clear_world_cache()
motion_gen.update_world(world_cfg)

```
_### Planning ###_ (**This doesn't find any solution.**)
```
plan_config = MotionGenPlanConfig(
    enable_opt=True,
    enable_graph=False,
    enable_graph_attempt=2,
    max_attempts=100,
    enable_finetune_trajopt=True,
    parallel_finetune=True,
)
result = motion_gen.plan_single(start_state, target_pose, plan_config)
plan = result.get_interpolated_plan()
```

**result returned was False**

But a good news is that if I change enable_graph to TRUE and enable_opt to FALSE and set interpolation_steps to 35 then planner does find the solution and performance number after fixing the way we update the world with obstacles is significantly better. Now with 7K obstacles being updated in the world; planning total time is 2-3seconds in comparison to 20-25seconds before. 
```
interpolation_steps = 35
plan_config = MotionGenPlanConfig(
    enable_opt=False,
    enable_graph=True,
    enable_graph_attempt=2,
    max_attempts=100,
    enable_finetune_trajopt=True,
    parallel_finetune=True,
)
result =motion_gen.plan_single(start_state, target_pose, plan_config)
plan = result.get_interpolated_plan()
```

**Reply by balakumar-s** [reply]
Can you share your robot yml file?

**Reply by sanjkar126** [reply]
Sharing the robot yml as txt file.

[robot_cfg.txt](https://github.com/NVlabs/curobo/files/14215919/robot_cfg.txt)

**Reply by balakumar-s** [reply]
I found at least one potential issue, we tuned the weights for max_acceleration: 15 and max_jerk: 500. 10x higher values will need re-tuning. Can you use these values and see if enable_opt works?

**Reply by sanjkar126** [reply]
Yes I did update the max_acc and max_jerk to 15 and 500 respectively. But still no luck if I enable_opt to TRUE.

**Reply by balakumar-s** [reply]
Can you share your robot's urdf and meshes so that we can try it out ?

**Reply by sanjkar126** [reply]
Balakumar, Is there a way to debug this on my end? Or enable additional logging in Curobo which I can then share with you.

**Reply by balakumar-s** [reply]
What are the velocity limits of your robot? Are they in the range of 2 radians? If they are very small (like 0.05 radians) that could also be an issue w.r.t. our smoothness weights.

**Reply by sanjkar126** [reply]
Have configured the velocity in URDF as for all the joints.
```
<limit
      lower="xxxxx"
      upper="yyyy"
      effort="3000"
      velocity="100" />
```

**Reply by balakumar-s** [reply]
Can you try adding `self_collision_check=False` to `MotionGenConfig.load_from_robot_config()`? and see if enable_opt succeeds?

**Reply by sanjkar126** [reply]
Tried setting self_collision_check=False and setting interpolation_steps=5000 with enable_opt=Ture & enable_graph=False. Still motion gen returns False.

**Reply by balakumar-s** [reply]
Can you also set `max_trajectory_dt=2.0` to `MotionGenConfig.load_from_robot_config()` and see if that helps?

**Reply by sanjkar126** [reply]
Set maximum_trajectory_dt=2.0 but still No Solution Found.

```
motion_gen_config = MotionGenConfig.load_from_robot_config(
            "my_robot.yml",
            "collision_table.yml",
            self.tensor_args,
            collision_checker_type=CollisionCheckerType.PRIMITIVE,
            num_trajopt_seeds=2, # 12 is default
            num_graph_seeds=2,
            use_cuda_graph=True,
            interpolation_dt=0.01,
            maximum_trajectory_time=1.0,
            collision_cache={"obb": n_obstacle_cuboids, "mesh": n_obstacle_mesh},
            trajopt_dt=trajopt_dt,
            trajopt_tsteps=trajopt_tsteps,
            trim_steps=trim_steps,
            collision_activation_distance=0.0,
            maximum_trajectory_dt=2.0,
        )
```

**Reply by sanjkar126** [reply]
> Bias node is not valid, not using bias node

Get this log line. Not sure if this indicates something

**Reply by balakumar-s** [reply]
That means that the robot is in collision at the retract_config. Try changing it so that there is no collision.

**Comment by Younch-15** [comment]
Current using v0.7.4 with Isaac Sim 4.0, facing exactly the same problem. When setting enable_graph as True, I always get FINETUNE_TRAJOPT_FAIL error no matter how I set enable_graph; when setting only enable_graph as True, the trajectory generated does not make any sense.

**Comment by hesic73** [comment]
I encountered the same problem. However, if I lock certain joints, the planning process succeeds.

**Comment by Arpitrf** [comment]
I have been facing a similar issue where motion gen would often return MotionGenStatus.TRAJOPT_FAIL. In my case I am trying to use motion planning for planning a mobile robot's base to a target pose (x, y, yaw). I found that if I increase the maximum_trajectory_dt to a vary high value (>60.0), I get better motion planning success rate. However, the success rate is still not high enough. @balakumar-s are there some other parameters that I could tune to improve the motion planning performance further? To add: I am testing the motion planning in a very simple env with no obstacles, so I am expecting the success rate to be close to 100%.

Another question related to setting  maximum_trajectory_dt to a vary high value (>60.0): What is the significance of this value? I am confused because when motion planning fails, it does not take 60 seconds (which is what I had expected).
